internal/http/services/owncloud/ocdav/tus.go

```diff
@@ -0,0 +1,140 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ocdav
+
+import (
+	"net/http"
+	"path"
+
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	typespb "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	tusd "github.com/tus/tusd/pkg/handler"
+)
+
+func (s *svc) handleTusPost(w http.ResponseWriter, r *http.Request, ns string) {
+	ctx := r.Context()
+	log := appctx.GetLogger(ctx)
+
+	w.Header().Add("Access-Control-Allow-Headers", "Tus-Resumable, Upload-Length, Upload-Metadata, If-Match")
+	w.Header().Add("Access-Control-Expose-Headers", "Tus-Resumable, Location")
+
+	// Test if the version sent by the client is supported
+	// GET methods are not checked since a browser may visit this URL and does
+	// not include this header. This request is not part of the specification.
+	if r.Header.Get("Tus-Resumable") != "1.0.0" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+	if r.Header.Get("Upload-Length") == "" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+
+	// read filename from metadata
+	meta := tusd.ParseMetadataHeader(r.Header.Get("Upload-Metadata"))
+	if meta["filename"] == "" {
+		w.WriteHeader(http.StatusPreconditionFailed)
+		return
+	}
+
+	// append filename to current dir
+	fn := path.Join(ns, r.URL.Path, meta["filename"])
+
+	// check tus headers?
+
+	// check if destination exists or is a file
+	client, err := s.getClient()
+	if err != nil {
+		log.Error().Err(err).Msg("error getting grpc client")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	sReq := &provider.StatRequest{
+		Ref: &provider.Reference{
+			Spec: &provider.Reference_Path{Path: fn},
+		},
+	}
+	sRes, err := client.Stat(ctx, sReq)
+	if err != nil {
+		log.Error().Err(err).Msg("error sending grpc stat request")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if sRes.Status.Code != rpc.Code_CODE_OK {
+		if sRes.Status.Code != rpc.Code_CODE_NOT_FOUND {
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	info := sRes.Info
+	if info != nil && info.Type != provider.ResourceType_RESOURCE_TYPE_FILE {
+		log.Warn().Msg("resource is not a file")
+		w.WriteHeader(http.StatusConflict)
+		return
+	}
+
+	if info != nil {
+		clientETag := r.Header.Get("If-Match")
```

Comment:
add TODO or raise ticket for "If-None-Match" header support as well

for example one can send a request with `If-None-Match: *` to prevent overwriting an already existing file in case of concurrency (saves doing a HEAD beforehand to check for existtence)

Reply:
tracked in https://github.com/cs3org/reva/issues/4